src/Workspaces/Core/Portable/Workspace/Solution/TextDocumentStates.cs

```diff
@@ -319,6 +319,24 @@ public async ValueTask<ChecksumsAndIds<DocumentId>> GetTextChecksumsAndIdsAsync(
         return new(documentChecksums, SelectAsArray(static s => s.Id));
     }
 
+    public async ValueTask<DocumentChecksumsAndIds> GetDocumentChecksumsAndIdsAsync(CancellationToken cancellationToken)
+    {
+        using var _1 = ArrayBuilder<Checksum>.GetInstance(_map.Count, out var attributeChecksums);
```

Comment:
>using var _1 = ArrayBuilder<Checksum>.GetInstance(_map.Count, out var attributeChecksums); [](http://example.com/codeflow?start=8&length=90)

nit: Since callstack is frequent, size is known and can be large enough to exceed ArrayBuilder's threshold, seems like just newing directly and using ImmutableCollectionMarkshal might be an interesting option

Reply:
(This probably holds true for the SelectAsArray methods too)

Reply:
done in followup.